packages/flutter_tools/lib/src/debug_adapters/error_formatter.dart

```diff
@@ -0,0 +1,187 @@
+    int indent = 0,
+    bool isError = false,
+    bool parseStackFrames = false,
+  }) {
+    if (text != null) {
+      final String indentString = '    ' * indent;
+      final String message = '$indentString${text.trim()}';
+
+      _BatchedOutput? output = batchedOutput.lastOrNull;
+      if (output == null || output.isError != isError || output.parseStackFrames != parseStackFrames) {
+        batchedOutput.add(output = _BatchedOutput(isError, parseStackFrames: parseStackFrames));
+      }
+      output.writeln(message);
+    }
+  }
+
+  /// Writes [node] to the output using [indent], recursing unless [recursive]
+  /// is `false`.
+  void _writeNode(_ErrorNode node, {int indent = 0, bool recursive = true}) {
+    // Errors, summaries and lines starting "Exception:" are marked as errors so
+    // they go to stderr instead of stdout (this may cause the client to colour
+    // them like errors).
+    final bool showAsError = node.level == _DiagnosticsNodeLevel.error ||
+        node.level == _DiagnosticsNodeLevel.summary ||
+        (node.description?.startsWith('Exception: ') ?? false);
+
+    if (node.showName && node.name != null) {
+      _write('${node.name}: ${node.description}', indent: indent, isError: showAsError);
+    } else if (node.description?.startsWith('#') ?? false) {
+      // Possible stack frame.
+      _write(node.description, indent: indent, isError: showAsError, parseStackFrames: true);
+    } else {
+      _write(node.description, indent: indent, isError: showAsError);
+    }
+
+    if (recursive) {
+      if (node.style != _DiagnosticsNodeStyle.flat) {
+        indent++;
+      }
+      _writeNodes(node.properties, indent: indent);
+      _writeNodes(node.children, indent: indent);
+    }
+  }
+
+  /// Writes [nodes] to the output.
+  void _writeNodes(List<_ErrorNode> nodes, {int indent = 0, bool recursive = true}) {
+    for (final _ErrorNode child in nodes) {
+      _writeNode(child, indent: indent, recursive: recursive);
+    }
+  }
+
+  /// Writes a simple summary of [node] to the output.
+  void _writeSummary(_ErrorNode node) {
+    final bool allChildrenAreLeaf = node.children.isNotEmpty &&
+        !node.children.any((_ErrorNode child) => child.children.isNotEmpty);
+    if (node.level == _DiagnosticsNodeLevel.summary || allChildrenAreLeaf) {
+      _writeNode(node, recursive: false);
+    }
+  }
+}
+
+/// A container for output to be sent to the client.
+///
+/// When multiple lines are being sent, they may be written to the same batch
+/// if the output options (error/stackFrame) are the same.
+class _BatchedOutput {
+  _BatchedOutput(this.isError, {this.parseStackFrames = false});
+
+  final bool isError;
+  final bool parseStackFrames;
+  final StringBuffer _buffer = StringBuffer();
+
+  String get output => _buffer.toString();
+
+  void writeln(String output) => _buffer.writeln(output);
+}
+
+enum _DiagnosticsNodeLevel {
+  error,
+  summary,
+}
+
+enum _DiagnosticsNodeStyle {
+  flat,
```

Comment:
are there going to be more values later?

Reply:
There are other values in the source enum, but this is the only value we currently use for this formatting, so I trimmed the enums just to the used values (otherwise we'd be needlessly keeping them in-sync with the source).

In future it would be nice to extract these classes/enums from `package:flutter` into a shared package that flutter_tools can use, but I failed to do that cleanly previously because the source types have lots of references to other Flutter framework types.